addons/udes_stock/tests/test_stock_reservation.py

```diff
@@ -95,3 +96,222 @@ def test_logs_debug_message_if_more_one_picking_processed(self):
         self.assertEqual(
             messages, ["Checking reservability for 1 pickings.", "Checking pickings 0-0"]
         )
+
+
+# Reserve Stock tests.
+
+# The principal conditions that affect stock reservation are whether a picking
+# is in a batch, and the values of the picking type's u_reserve_batches and
+# u_handle_partial flags. The stock level conditions that affect reservation
+# are: sufficent stock, no stock, some lines are not reservable and some lines
+# are partially available.
+
+# To test this, we have a matrix of configuration options, initial stock levels
+# and expected reservations. The ReservationMixin class has tests for each of
+# the stock level conditions; the TestCase classes provide the configuration
+# for each test method.
+
+# Test configuration, stock levels and expected outputs, keyed on:
+# Pickings are in batches, PickingType.u_reserve_batches, PickingType.u_handle_partials.
+# Values order: all available, none available, partially available, line partially available
+# In-values order:
+# quantity in stock, expected reservation
+# each picking has two identical lines (2 x 10)
+# each batch has two identical pickings
+RESERVATION_MATRIX = {
+    (True, True, True): [(40, 40), (0, 0), (30, 30), (35, 35)],
```

Comment:
Seems like could use binary in mapping 7,6,5,4,3,2,1,0 and convert to binary , maybe no need for classes but a loop in range (8)